### PR TITLE
add firmware check for duplicate versions

### PR DESCRIPTION
**Pull request by ozym**
One of the complexities of generating GNSS processing configuration files is handling firmware changes, most formats require a break or a new entry when a change has been made.

However, in some cases no change was made, just an added firmware entry. This results in multiple false firmware changes which need to be cleaned up. 

This PR hopefully removes this last step, it adds a test to check that the firmware has actually changed, and updates the current firmware list to pass the test.

### CODE REVIEW

**Comment by elidana** [comment]
Hi @ozym , that is great and it is nice to see unnecessary entries removed from delta.
A question I have (while we do the real work to review this pr).
The question is only relevant while we transition from old to new sitelog provision: will this change anything in the creation of the rinexxml files, that the old python codes are using?
(I can check myself...with some good uninterrupted time! :D )

File: install/firmware.csv [hunk 1]
```diff
@@ -1710,10 +1686,7 @@ Trimble Navigation Ltd.,TRIMBLE NETRS,4503243993,1.0-3,2004-12-01T08:00:00Z,2006
 Trimble Navigation Ltd.,TRIMBLE NETRS,4503243993,1.1-5,2006-04-18T00:00:01Z,2008-12-10T14:37:08Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4503243993,1.2-0,2008-12-10T14:37:09Z,9999-01-01T00:00:00Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.0-3,2004-12-01T08:00:00Z,2006-05-22T10:11:05Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-3,2006-05-22T10:11:05Z,2006-05-22T15:47:42Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-3,2006-05-22T15:47:42Z,2006-05-23T14:04:05Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-3,2006-05-23T14:04:05Z,2006-05-25T14:48:22Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-3,2006-05-25T14:48:22Z,2006-10-06T02:26:41Z,
+Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-3,2006-05-22T10:11:05Z,2006-10-06T02:26:41Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-5,2006-10-06T02:26:45Z,2008-09-22T10:30:47Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.2-0,2008-09-22T10:30:48Z,9999-01-01T00:00:00Z,
```
Comment (elidana) [inline install/firmware.csv:1691]:
```
[elidana@hutl16534 tmp.378853]$ teqc +meta WHNG200602130104a.dat 
start date & time:       2006-02-13 01:04:30.000
receiver ID number:      4503244058
receiver type:           TRIMBLE NETRS
receiver firmware:       1.1-3
```
```
! Alert ! 'WHNG200610050136a.dat': change of receiver firmware from '1.13' to '1.15' ~ 2006 Oct  4 23:59:30.000
! Alert ! 'GISB200809212300a.dat': change of receiver firmware from '1.15' to '1.20' ~ 2008 Sep 21 22:59:30.000
```

propose to change these entries to:
```
Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.0-3,2004-12-01T08:00:00Z,2006-02-13T00:00:00Z,
Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-3,2006-02-13T00:00:01Z,2006-10-04T23:59:29Z,
Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.1-5,2006-10-04T23:59:30Z,2008-09-21T22:59:29Z,
Trimble Navigation Ltd.,TRIMBLE NETRS,4503244058,1.2-0,2008-09-21T22:59:30Z,9999-01-01T00:00:00Z,
```

File: install/firmware.csv [hunk 2]
```diff
@@ -1728,17 +1701,11 @@ Trimble Navigation Ltd.,TRIMBLE NETRS,4632120756,1.1-5,2006-10-11T00:00:00Z,2008
 Trimble Navigation Ltd.,TRIMBLE NETRS,4632120756,1.2-0,2008-12-05T14:54:46Z,9999-01-01T00:00:00Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4635120781,1.1-5,2006-10-11T00:00:00Z,2009-04-07T15:01:45Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4635120781,1.2-0,2009-04-07T15:01:46Z,9999-01-01T00:00:00Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.1-5,2006-10-11T00:00:00Z,2008-12-10T14:00:16Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.1-5,2008-12-10T14:00:17Z,2008-12-10T14:41:49Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.1-5,2008-12-10T14:41:50Z,2008-12-11T12:01:24Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.1-5,2008-12-11T12:01:25Z,2009-04-07T18:13:14Z,
+Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.1-5,2006-10-11T00:00:00Z,2009-04-07T18:13:14Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.2-0,2009-04-07T18:13:15Z,9999-01-01T00:00:00Z,
```
Comment (elidana) [inline install/firmware.csv:1705]: `! Alert ! 'KARA200904070700a.dat': change of receiver firmware from '1.15' to '1.20' ~ 2009 Apr  7 06:59:30.000`

propose to change to:
```
Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.1-5,2006-10-11T00:00:00Z,2009-04-07T06:59:29Z,
Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.2-0,2009-04-07T06:59:30Z,9999-01-01T00:00:00Z,
```

**Comment by ozym** [comment]
@elidana I have updated the firmware file as above, I didn't make changes to the existing changes for 4415233388 and 4635120795

Again could you check I've made the correct changes, this looks like a much better approach.

File: install/firmware.csv [hunk 3]
```diff
@@ -1449,8 +1449,7 @@ Trimble Navigation Ltd.,TRIMBLE NETRS,4352230457,1.1-5,2006-04-04T11:37:02Z,9999
 Trimble Navigation Ltd.,TRIMBLE NETRS,4356263016,1.0-2,2003-12-31T00:00:00Z,9999-01-01T00:00:00Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.0-3,2003-12-26T08:00:00Z,2006-04-04T15:30:43Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.1-3,2006-04-04T15:30:43Z,2006-04-05T09:57:05Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.1-5,2006-04-05T09:57:06Z,2008-10-17T15:09:46Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.1-5,2008-10-17T15:09:47Z,9999-01-01T00:00:00Z,
+Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.1-5,2006-04-05T09:57:06Z,9999-01-01T00:00:00Z,
```
Comment (elidana) [inline install/firmware.csv:1452]: metadata from data
```
! Alert ! 'KAPT200604040346a.dat': change of receiver firmware from '1.03' to '1.13' ~ 2006 Apr  4 02:59:30.000
! Alert ! 'KAPT200604042200a.dat': change of receiver firmware from '1.13' to '1.15' ~ 2006 Apr  4 21:59:30.000
```

proposed changes:
```
Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.0-3,2003-12-26T08:00:00Z,2006-04-04T02:59:29Z,
Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.1-3,2006-04-04T02:59:30Z,2006-04-04T21:59:29Z,
Trimble Navigation Ltd.,TRIMBLE NETRS,4415233388,1.1-5,2006-04-04T21:59:30Z,9999-01-01T00:00:00Z,
```

File: install/firmware.csv [hunk 4]
```diff
@@ -1734,17 +1706,11 @@ Trimble Navigation Ltd.,TRIMBLE NETRS,4632120756,1.1-5,2006-10-11T00:00:00Z,2008
...
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.2-0,2009-04-07T18:13:15Z,9999-01-01T00:00:00Z,
+Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.1-5,2006-10-11T00:00:00Z,2009-04-07T06:59:29Z,
+Trimble Navigation Ltd.,TRIMBLE NETRS,4635120785,1.2-0,2009-04-07T06:59:30Z,9999-01-01T00:00:00Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4635120794,1.1-5,2006-10-11T00:00:00Z,2009-04-08T10:23:06Z,
 Trimble Navigation Ltd.,TRIMBLE NETRS,4635120794,1.2-0,2009-04-08T10:23:07Z,9999-01-01T00:00:00Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120795,1.1-5,2006-10-11T00:00:00Z,2008-12-10T16:04:39Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120795,1.1-5,2008-12-10T16:04:40Z,2009-04-08T10:29:56Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120795,1.1-5,2009-04-08T10:29:57Z,2009-04-08T10:47:30Z,
-Trimble Navigation Ltd.,TRIMBLE NETRS,4635120795,1.1-5,2009-04-08T10:47:31Z,9999-01-01T00:00:00Z,
+Trimble Navigation Ltd.,TRIMBLE NETRS,4635120795,1.1-5,2006-10-11T00:00:00Z,9999-01-01T00:00:00Z,
```
Comment (elidana) [inline install/firmware.csv:1713]: this one looks good, is a single firmware version that remained the same. Receiver was installed at two different sites, and some randomly sampled days are all showing firmware 1.1-5 was installed, so no need to change this updated entry (2006 to 9999 for 1.1-5)

**Pull request merged**